Improve module checkpointing performance

TOTBWF opened the pull request:
I was doing some profiling of agda against `1lab` builds the other day, and noticed that we were spending upwards of 5% of our time inside of the `Ord` instance of `Abstract.Name`. After some further digging, I noticed that [checkpoint](https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/Monad/Context.hs#L101) unions `stModuleCheckpoints` with a potentially extended version of itself every time we extend the context: for moderately sized files, this can happen hundreds of thousands of times!

This PR replaces the `stModuleCheckpoints` map with a stack containing sets of module names and checkpoint ids: this turns the expensive call to `union` into a single integer comparison in the happy path, and reduces that 5% to 0.1%.

TOTBWF committed:
[perf] add UNPACK pragma to NameId in Abstract.Name

In my benchmarking, this shaves about 11 seconds off of a clean build
of the 1lab.

TOTBWF committed:
[debug] Add verbose print for difference between module checkpoints

TOTBWF committed:
[perf] Replace module checkpoint map with a stack

This reduces runtime by about 5% in my benchmarks, as we get
to avoid doing repeated calls to `Map.union` inside of `checkpoint`.

TOTBWF committed:
[fix] Fix wayward comment

TOTBWF committed:
[fix] dont unwind past the checkpoint id we are looking for

jespercockx commented:
Oh wow, that looks like a really significant optimization basically for free. Thanks for the PR!

I believe @liesnikov had also noticed the extremely high number of calls to compare on abstract names when counting internal function calls while typechecking the standard library, but we never really got to the bottom of that. Do you have some numbers on the timings on the standard library with and without this PR?

TOTBWF commented:
Thanks! Unfortunately the CI failure is indicative of a real bug in my code, so I can't load the stdlib right now 😁 Going to bite the bullet and do what I should've done in the first place, and add the old code alongside the new so that we can assert to find inconsistencies. 

Also, the biggest speedups will probably be in cubical code: checking an hcomp system requires a quite a few context extensions!

TOTBWF commented:
There's also some other nice low-hanging fruit in this area of the code. Using a `Map` for checkpoints in TCEnv is seems somewhat suboptimal: `IntMap` should be a drop-in replacement. However, I have a hunch that a linked list might actually be better here: we extend the context a lot more often than we look up checkpoints, and rebalancing a map makes us lose some possible sharing of the spine. Once I iron out the kinks in this PR I'll do some experiments.

TOTBWF committed:
[fix] Fix ordering in Pretty instance of ModuleCheckpoints

TOTBWF committed:
[fix] More precise module stack unwinding code

TOTBWF committed:
[fix] Fix checkpoint unwinding after TCState resets

TOTBWF commented:
Looks like the bug was that something reset the `TCState` while inside of a `checkpoint` to a state that had no module checkpoints. I've made `checkpoint` more robust to state resets, but it would probably be good to track down the culprit.

TOTBWF commented:
Some timings: the master build of agda takes [418.785 seconds](https://github.com/agda/agda/actions/runs/15653547589/job/44101780101#step:7:1165) to typecheck the standard library in CI. After these changes, we are down to [402.391 seconds](https://github.com/agda/agda/actions/runs/15765385635/job/44441241555?pr=7951#step:7:1165)!

TOTBWF committed:
[test] Add test for interleaving modules inside records and fields

This test is a minimal reproducer of the structure of `Cat.Bi.Base`
from the 1lab, which fails to typecheck with the optimized module
checkpointing code.

TOTBWF commented:
Just tested this against a clean `1lab` build, and we save 100 seconds on my machine. Some stats:
```
-- master
Total                                   1,315,318ms            
Miscellaneous                               3,903ms            
Typing                                    132,030ms (796,698ms)
Typing.CheckRHS                           215,773ms            
Typing.InstanceSearch                       5,692ms (155,040ms)
Typing.InstanceSearch.FilterCandidates     89,915ms            
Typing.InstanceSearch.CheckOverlap         37,781ms            
Typing.InstanceSearch.InitialCandidates    21,646ms            
Typing.Reflection                         115,873ms            
Typing.OccursCheck                         69,656ms            
Typing.TypeSig                             39,697ms            
Typing.CheckLHS                            29,207ms  (33,931ms)
Typing.CheckLHS.UnifyIndices                4,723ms            
Typing.Generalize                          32,791ms            
Typing.With                                 1,904ms            
Serialization                             105,592ms (125,322ms)
Serialization.BinaryEncode                  8,394ms            
Serialization.Compress                      8,090ms            
Serialization.Sort                          2,540ms            
Serialization.BuildInterface                  704ms            
Positivity                                101,260ms            
Parsing                                    12,286ms  (78,203ms)
Parsing.OperatorsExpr                      59,115ms            
Parsing.OperatorsPattern                    6,801ms            
Scoping                                     7,051ms  (52,352ms)
Scoping.InverseScopeLookup                 45,301ms            
Deserialization                            26,277ms  (36,225ms)
Deserialization.Compaction                  9,948ms            
DeadCode                                      497ms  (35,984ms)
DeadCode.DeadCodeReachable                 35,487ms            
InterfaceInstantiateFull                   31,945ms            
Coverage                                   25,605ms  (30,730ms)
Coverage.UnifyIndices                       5,125ms            
Termination                                 3,765ms  (12,501ms)
Termination.RecCheck                        7,777ms            
Termination.Compare                           848ms            
Termination.Graph                             109ms            
Highlighting                                6,709ms            
Import                                      1,602ms            
ProjectionLikeness                          1,142ms            
Injectivity                                   701ms            
ModuleName                                     37ms
```

```
-- 209f1e0
Total                                   1,210,243ms            
Miscellaneous                               4,016ms            
Typing                                    113,881ms (712,710ms)
Typing.CheckRHS                           184,642ms            
Typing.InstanceSearch                       6,176ms (131,080ms)
Typing.InstanceSearch.FilterCandidates     77,475ms            
Typing.InstanceSearch.CheckOverlap         25,360ms            
Typing.InstanceSearch.InitialCandidates    22,063ms            
Typing.Reflection                         109,684ms            
Typing.OccursCheck                         89,305ms            
Typing.TypeSig                             30,194ms            
Typing.CheckLHS                            23,317ms  (27,006ms)
Typing.CheckLHS.UnifyIndices                3,688ms            
Typing.Generalize                          25,298ms            
Typing.With                                 1,617ms            
Serialization                              96,088ms (121,836ms)
Serialization.BinaryEncode                 10,087ms            
Serialization.Compress                      8,988ms            
Serialization.Sort                          5,945ms            
Serialization.BuildInterface                  725ms            
Parsing                                    12,518ms  (81,784ms)
Parsing.OperatorsExpr                      62,359ms            
Parsing.OperatorsPattern                    6,905ms            
Positivity                                 75,372ms            
Scoping                                     7,060ms  (52,171ms)
Scoping.InverseScopeLookup                 45,111ms            
Deserialization                            29,550ms  (40,186ms)
Deserialization.Compaction                 10,636ms            
DeadCode                                      498ms  (40,068ms)
DeadCode.DeadCodeReachable                 39,570ms            
InterfaceInstantiateFull                   30,586ms            
Coverage                                   24,170ms  (28,217ms)
Coverage.UnifyIndices                       4,047ms            
Termination                                 4,652ms  (12,922ms)
Termination.RecCheck                        7,160ms            
Termination.Compare                           997ms            
Termination.Graph                             111ms            
Highlighting                                6,707ms            
Import                                      1,601ms            
ProjectionLikeness                          1,400ms            
Injectivity                                   628ms            
ModuleName                                     37ms
```

There are some other optimizations I'd like to make, but they are independent of this PR, so this should be ready for review.

andreasabel commented:
@Mergifyio rebase
